### PR TITLE
Implemented hierarchy cache - delete has problems

### PR DESCRIPTION
**Pull request by tobyryuk**
Hierarchy cached added, only thing that’s having problems is the removal of items.

### CODE REVIEW

File: src/MemcachedCachePool.php [hunk 1]
```diff
@@ -45,7 +49,11 @@ protected function clearAllObjectsFromCache()
 
     protected function clearOneObjectFromCache($key)
     {
-        if ($this->cache->delete($key)) {
+        $keyString = $this->getHierarchyKey($key, $path);
```
Comment (Nyholm) [inline src/MemcachedCachePool.php]: I think you have forgotten to do `$this->commit()` to make sure to move all your items in the deferred queue to the Memcached storage.

**Comment by tobyryuk** [comment]
Added 

``` php
$this->commit();
```

Same results.

File: src/MemcachedCachePool.php [hunk 2]
```diff
@@ -53,12 +62,15 @@ protected function clearOneObjectFromCache($key)
...
-    protected function storeItemInCache($key, CacheItemInterface $item, $ttl)
+    protected function storeItemInCache($key, CacheItemInterface $item, $ttl = -1)
     {
-        if ($ttl === null) {
-            $ttl = 0;
-        }
+        $key = $this->getHierarchyKey($key);
 
-        return $this->cache->set($key, $item, $ttl);
+        return $this->cache->set($key, $item, 0, $ttl);
```
Comment (Nyholm) [inline src/MemcachedCachePool.php]: I think it slipped in an error here. Memcached::set() has only three parameters. 

Don't we have to make sure `$ttl` is not null?

Comment (Nyholm) [inline src/MemcachedCachePool.php]: See https://github.com/php-cache/adapter-common/blob/master/src/AbstractCachePool.php#L182

**Comment by Nyholm** [comment]
Awesome, Thank you for this PR.

**Comment by Nyholm** [comment]
Hm, I think you have to debug this. Start with investigate return value of `$this->cache->increment($key)`.  Se if you can find out what happens if you try increment on a stored null etc.

**Comment by tobyryuk** [comment]
When incrementing $path in the Redis adapter, it returns an integer while here it returns a bool.

**Comment by Nyholm** [comment]
> When incrementing $path in the Redis adapter, it returns an integer while here it returns a bool.

A `bool` or `false`? Have a look here: http://php.net/manual/en/memcached.increment.php

**Comment by tobyryuk** [comment]
It returns false.

**Comment by tobyryuk** [comment]
Now it's only the tags that are in trouble.

**Comment by Nyholm** [comment]
What does it mean when you do `$this->cache->setOption(\Memcached::OPT_BINARY_PROTOCOL,true);`? How does this effect all other calls with this Memcached class?

**Comment by tobyryuk** [comment]
I'm not 100% sure of how it effects the other calls, but it allows the $path to be incremented even though it's a string containing non numeric characters as well.

**Comment by Nyholm** [comment]
> I'm not 100% sure of how it effects the other calls,

Hm, so it could have strange side-effects. How about being safe and use a `get()` and then `set()`?

**Comment by tobyryuk** [comment]
Found [this](http://www.slideshare.net/tmaesaka/memcached-binary-protocol-in-a-nutshell-presentation), according to that, there should not be side effects to using the binary protocol.

**Comment by Nyholm** [comment]
Okey, cool. It seams like it is more performant as well. But I will suggest that you put the setOption in the constructor

**Comment by tobyryuk** [comment]
Nice timing, just moved it.

**Comment by tobyryuk** [comment]
I think my expertise in caching ends here, from what I've gathered browsing the web, adding tags to Memcached is not the easiest thing to do.

**Comment by Nyholm** [comment]
> adding tags to Memcached is not the easiest thing to do.

It sure isn't easy. 

I've debugged this and it seams like you are doing every thing correct. It is a limitation in Memcahced that says the key should not be more than 250 chars. I think we need to do some modification in the tagging repository. 

Can you squash your commits so I can merge this when tagging is ready?

**Comment by Nyholm** [comment]
Also, bump `cache/taggable-cache` to `^0.4`.

**Comment by tobyryuk** [comment]
I think I managed to squash it, not too used with squashing.

**Comment by Nyholm** [comment]
Awesome, yeah I looks like you've done everything correct. Thank you!

**Comment by tobyryuk** [comment]
Phew, great!

**Comment by Nyholm** [comment]
Ping @aequasi, what do you think about this? 
We fear collision with the change of 10^-45, which is rare... very rare. Should we implement a `sha1()` in the `getHierarchalKey` instead?

File: src/MemcachedCachePool.php [hunk 3]
```diff
@@ -59,6 +69,28 @@ protected function storeItemInCache($key, CacheItemInterface $item, $ttl)
...
+     *
+     * @param $key
+     * @param null $ref
+     */
+    private function trimKey($key)
+    {
+        if (strlen($key) < 250) {
+            return $key;
+        }
+        // This should maybe be logged
```
Comment (cryptiklemur) [inline src/MemcachedCachePool.php:93]: Put a new line above this comment

**Pull request merged**